src/nodedef.cpp

```diff
@@ -283,7 +283,20 @@ void TileDef::deSerialize(std::istream &is, NodeDrawType drawtype, u16 protocol_
 void TextureSettings::readSettings()
 {
 	connected_glass                = g_settings->getBool("connected_glass");
-	opaque_water                   = g_settings->getBool("opaque_water");
+
+	//Update from opaque_water to translucent_water
```

Comment:
This can be simplified to

```c++
if (g_settings->exists("opaque_water"))
    g_settings->setDefault("translucent_water", g_settings->getBool("opaque_water") ? "false" : "true");
translucent_water = g_settings->getBool("translucent_water");
```

However, I'm not very sold on changing the global setting defaults here. Why not instead do the following:

```c++
// Fall back to old setting with inverted meaning if new setting is absent
if (g_settings->exists("opaque_water") && !g_settings->exists("translucent_water")) {
    translucent_water = !g_settings->getBool("opaque_water");
} else {
    translucent_water = g_settings->getBool("translucent_water");
}
```

This would avoid mutating `g_settings`, hence staying true to the name of `readSettings`.

Reply:
I tried that origonaly, but it seemed conflicted with the default setting. I can try again wit your code. I do agree that would be the best way though.

Reply:
Just tested that does NOT work because it conflicts with the defined default so that no matter what `g_settings->exists("translucent_water")` does exist thus will never run.

Reply:
My bad, `exists` also checks for defaults. We could use `existsLocal`, but that would only check a single layer.

I think one way to solve it would be to introduce a method `Settings *Settings::getTopLayer(const std::string &setting)` which returns you a pointer to the top layer on which a setting was found, or `nullptr`. This could then be used via `g_settings->getTopLayer("translucent_water") != g_settings->getLayer(SL_DEFAULTS)`.

This should work (untested):
```diff
diff --git a/src/settings.cpp b/src/settings.cpp
index 74d00762d..2eadede52 100644
--- a/src/settings.cpp
+++ b/src/settings.cpp
@@ -657,14 +657,18 @@ bool Settings::getNoiseParamsFromGroup(const std::string &name,
 	return true;
 }
 
-
-bool Settings::exists(const std::string &name) const
+Settings *Settings::getTopLayer(const std::string &name) const
 {
 	if (existsLocal(name))
-		return true;
+		return this;
 	if (auto parent = getParent())
-		return parent->exists(name);
-	return false;
+		return parent->getTopLayer(name);
+	return nullptr;
+}
+
+bool Settings::exists(const std::string &name) const
+{
+	return getTopLayer(name) != nullptr;
 }
 
 
diff --git a/src/settings.h b/src/settings.h
index 692001c59..5d44fa158 100644
--- a/src/settings.h
+++ b/src/settings.h
@@ -174,6 +174,8 @@ class Settings {
 
 	// return all keys used in this object
 	std::vector<std::string> getNames() const;
+	// get the top layer on which a setting exists, or nullptr if it does not exist
+	Settings *getTopLayer(const std::string &name) const;
 	// check if setting exists anywhere in the hierarchy
 	bool exists(const std::string &name) const;
 	// check if setting exists in this object ("locally")
```